Instruct PLEG to detect pod sandbox state changes

yujuhong opened the pull request:
This PR adds a Sandboxes list in `kubecontainer.Pod`, so that PLEG can check
sandbox changes using `GetPods()` . The sandboxes are treated as regular
containers (type `kubecontainer.Container`) for now to avoid additional
changes in PLEG.

/cc @feiskyer @yifan-gu @euank

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/31395)

<!-- Reviewable:end -->

yujuhong committed:
Add "Sandboxes" to the kubecontainer.Pod

This field will only be used by the kuberuntime package and should be ignored
by other type of runtimes.

yujuhong committed:
Extend PLEG to handle pod sandboxes

PLEG will treat them as if they are regular containers and detect changes the
same manner. Note that this makes an assumption that container IDs will not
collide with the podsandbox IDs.

yifan-gu commented on pkg/kubelet/kuberuntime/helpers.go:
```diff
@@ -94,6 +109,21 @@ func (m *kubeGenericRuntimeManager) toKubeContainer(c *runtimeApi.Container) (*k
+// kubecontainer.Containers to avoid substantial changes to PLEG.
+// TODO: Remove this once it becomes obsolete.
+func (m *kubeGenericRuntimeManager) sandboxToKubeContainer(s *runtimeApi.PodSandbox) (*kubecontainer.Container, error) {
+	if s == nil || s.Id == nil || s.State == nil {
```
Why these could be nil?

yifan-gu commented on pkg/kubelet/pleg/generic.go:
```diff
@@ -291,6 +292,18 @@ func getContainersFromPods(pods ...*kubecontainer.Pod) []*kubecontainer.Containe
+			if cidSet.Has(cid) {
+				continue
+			}
+			glog.Errorf("sandbox %+v", c)
```
what does this log message try to tell?

yifan-gu commented on pkg/kubelet/pleg/generic_test.go:
```diff
@@ -430,3 +430,69 @@ func TestRelistWithReinspection(t *testing.T) {
+	actual := getEventsFromChannel(ch)
+	verifyEvents(t, expected, actual)
+
+	// The second relist should not send out any event because no container
```
no containers are changed.

yifan-gu commented:
LGTM after some nits.

yujuhong commented:
Review status: 0 of 6 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/kubelet/kuberuntime/helpers.go, line 117 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31395#-KPz24DNz2YzSAbTbC1R-r2-117:-KPzLqDrpt5X708hn_xa:b8i5oif) ([raw file](https://github.com/kubernetes/kubernetes/blob/d796a9d1e6b4506adc604257d4fe78e55bf85d90/pkg/kubelet/kuberuntime/helpers.go#L117)):*
<details><summary><i>Previously, yifan-gu (Yifan Gu) wrote…</i></summary>
> Why these could be nil?
</details>
Don't want to panic when there the runtime misbehaves ;)

---

*[pkg/kubelet/pleg/generic.go, line 302 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31395#-KPz24DQImam8lp2kOk8-r2-302:-KPzMAd2bTX-oXGgVLWT:b-2koow1) ([raw file](https://github.com/kubernetes/kubernetes/blob/d796a9d1e6b4506adc604257d4fe78e55bf85d90/pkg/kubelet/pleg/generic.go#L302)):*
<details><summary><i>Previously, yifan-gu (Yifan Gu) wrote…</i></summary>
> what does this log message try to tell?
</details>
You've got me. It was for debugging only. Removed.

---

*[pkg/kubelet/pleg/generic_test.go, line 466 \[r2\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31395#-KPz24DQImam8lp2kOk9-r2-466:-KPzMLg5UnXG2KxyMEkD:bb6un1j) ([raw file](https://github.com/kubernetes/kubernetes/blob/d796a9d1e6b4506adc604257d4fe78e55bf85d90/pkg/kubelet/pleg/generic_test.go#L466)):*
<details><summary><i>Previously, yifan-gu (Yifan Gu) wrote…</i></summary>
> no containers is changed.
</details>
Done

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31395)*
<!-- Sent from Reviewable.io -->

yujuhong commented:
Thanks for the review! Addressed the comments.  
I didn't create a new commit since the changes were trivial. Hope that's ok.

-release-note-label-needed 
+release-note-none 

---







*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31395#-:-KPzMi6fcPTzH5_fahye:b-xkrnmt)*
<!-- Sent from Reviewable.io -->

feiskyer commented:
LGTM.

feiskyer commented:
@yujuhong Could we make one of #30267 and this PR in so that following PRs don't need to depend on two different branches?

yujuhong commented:
> @yujuhong Could we make one of #30267 and this PR in so that following PRs don't need to depend on two different branches?

Both these two PRs have been reviewed, so probably won't require a lot of rebasing. I think you can just cherry-pick the commits while we're waiting for the release branch to be cut...

feiskyer commented:
@yujuhong OK, will cherry-pick this PR in following PRs.

euank commented:
Reviewed 4 of 4 files at r1, 1 of 3 files at r2, 2 of 2 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/kubelet/kuberuntime/kuberuntime_manager.go, line 237 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31395#-KQ7ZlgJBtklzkJ8r3FR:-KQ7ZlgKBfqquY8qZU8O:b-3jxe0u) ([raw file](https://github.com/kubernetes/kubernetes/blob/1081c74b57833744f76c967fb6f12449f79f4d6a/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L237)):*
> ```Go
> 			pods[podUID] = &kubecontainer.Pod{
> 				ID:        podUID,
> 				Name:      s.Metadata.GetName(),
> ```

Above you have a defensive `s == nil` check in `sandboxToKubeContainer`, but here you dereference `s` anyways before passing it into the function, so we know it won't be nil there.

If there's a good reason to be super defensive, sure, but as is it's inconsistently defensive and so isn't really guarding against anything.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31395)*
<!-- Sent from Reviewable.io -->

yujuhong commented:
*[pkg/kubelet/kuberuntime/kuberuntime_manager.go, line 237 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31395#-KQ7ZlgJBtklzkJ8r3FR:-KQ9HHUYylzjBCzjP__Y:b-5rw15m) ([raw file](https://github.com/kubernetes/kubernetes/blob/1081c74b57833744f76c967fb6f12449f79f4d6a/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L237)):*
<details><summary><i>Previously, euank (Euan Kemp) wrote…</i></summary>
> Above you have a defensive `s == nil` check in `sandboxToKubeContainer`, but here you dereference `s` anyways before passing it into the function, so we know it won't be nil there.
> 
> If there's a good reason to be super defensive, sure, but as is it's inconsistently defensive and so isn't really guarding against anything.
</details>
Good point. I think I just moved the original code without modifying too much. Will fix this.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31395)*
<!-- Sent from Reviewable.io -->

yujuhong commented:
@euank PTAL, thanks!

euank commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/kubelet/kuberuntime/kuberuntime_manager.go, line 237 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31395#-KQ7ZlgJBtklzkJ8r3FR:-KQMRRjjFINqJ4FWvy_V:b-8zwuts) ([raw file](https://github.com/kubernetes/kubernetes/blob/1081c74b57833744f76c967fb6f12449f79f4d6a/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L237)):*
<details><summary><i>Previously, yujuhong (Yu-Ju Hong) wrote…</i></summary>
> Good point. I think I just moved the original code without modifying too much. Will fix this.
</details>
My point was that you dereference `s` not `s.Metadata`. The inconsistency remains because you now dereference `s` in the if statement you added, and then still check for `s == nil` one step further down the call-stack.

Since it's a private function anyways, I think it makes sense to just make `s` being non-nil a precondition.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31395)*
<!-- Sent from Reviewable.io -->

yujuhong commented:
*[pkg/kubelet/kuberuntime/kuberuntime_manager.go, line 237 \[r3\]](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31395#-KQ7ZlgJBtklzkJ8r3FR:-KQMrbvodogQap59tKvQ:b-813s0u) ([raw file](https://github.com/kubernetes/kubernetes/blob/1081c74b57833744f76c967fb6f12449f79f4d6a/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L237)):*
<details><summary><i>Previously, euank (Euan Kemp) wrote…</i></summary>
> My point was that you dereference `s` not `s.Metadata`. The inconsistency remains because you now dereference `s` in the if statement you added, and then still check for `s == nil` one step further down the call-stack.
> 
> Since it's a private function anyways, I think it makes sense to just make `s` being non-nil a precondition.
</details>
The `sandboxToKubeContainer` function doesn't use the `s.Metadata`, so it checks only `s`.  It is a private function, but it can be used in multiple places. I don't think doing the checks is completely redundant.

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/kubernetes/31395)*
<!-- Sent from Reviewable.io -->

yujuhong commented:
Squashed the commits. @euank are you ok with the PR?

euank commented:
I think it's overly defensive, but no problem with it. LGTM.

Random-Liu commented:
Change milestone to get this in queue. @feiskyer

feiskyer commented:
@Random-Liu Thanks a lot.
